File: sdks/typescript/src/presence.ts
import { StarfishError, type StarfishFrame } from "./types.js";
import type { Connection } from "./connection.js";
import type { Session } from "./session.js";
import { nextId } from "./id.js";
import { Observable } from "./emitter.js";
import { MAX_PRESENCE_SIZE, validatePayloadSize } from "./limits.js";
import { validateSerializable } from "./validate.js";

export class Presence {
  private connection: Connection;
  private session: Session;
  private presenceMap = new Map<string, any>();

  readonly presence$ = new Observable<Map<string, any>>(new Map());

  constructor(connection: Connection, session: Session) {
    this.connection = connection;
    this.session = session;
  }

  set(payload: any): void {
    const sessionName = this.requireSession();
    validateSerializable(payload, "Presence payload");
    const json = JSON.stringify(payload);
    validatePayloadSize(json, MAX_PRESENCE_SIZE, "Presence payload");

    const frame: StarfishFrame = {
      header: {
        id: nextId("pres"),
        resource: "presence",
        method: "set",
        kind: "request",
        session: sessionName,
      },
      payload,
    };

    this.connection.send(frame);
  }

  handleFrame(frame: StarfishFrame): void {
    if (
      frame.header.resource === "presence" &&
      frame.header.method === "updated" &&
      frame.header.from
    ) {
      this.presenceMap.set(frame.header.from, frame.payload);
      this.presence$.set(new Map(this.presenceMap));
    }
  }

  clear(): void {
    this.presenceMap.clear();
    this.presence$.set(new Map());
  }

  private requireSession(): string {
    const session = this.session.current;
    if (!session) {
      throw new StarfishError("NO_SESSION", "Not in a session. Call join() first.");
    }
    return session;
  }
}

File: sdks/swift/Sources/StarfishClient/Presence.swift
import Foundation

/// Manages client presence: setting own presence and observing peers.
public final class Presence: @unchecked Sendable {
    private let connection: Connection
    private let session: Session
    private var presenceMap: [String: AnyCodable] = [:]

    public let presence$ = Observable<[String: AnyCodable]>([:])

    init(connection: Connection, session: Session) {
        self.connection = connection
        self.session = session
    }

    public func set(_ payload: AnyCodable) throws {
        let sessionName = try session.require()

        let json = try encodePayload(payload)
        try validatePayloadSize(json, limit: Limits.maxPresenceSize, label: "Presence payload")

        let frame = StarfishFrame(
            header: StarfishHeader(
                id: connection.idGen.nextId(prefix: "pres"),
                resource: "presence",
                method: "set",
                kind: .request,
                session: sessionName
            ),
            payload: ["data": payload]
        )

        try connection.send(frame)
    }

    func handleFrame(_ frame: StarfishFrame) {
        guard frame.header.resource == "presence" && frame.header.method == "updated" else { return }
        guard let from = frame.header.from else { return }
        if let data = frame.payload?["data"] {
            presenceMap[from] = data
        } else {
            presenceMap[from] = AnyCodable(frame.payload as Any)
        }
        presence$.set(presenceMap)
    }

    func clear() {
        presenceMap.removeAll()
        presence$.set([:])
    }
}
